src/Stack/Types/Config.hs

```diff
@@ -1149,8 +1164,37 @@ instance Show ConfigException where
         [ "Invalid ghc-variant value: "
         , v
         ]
+    show (BadStackRootEnvVar envStackRoot) = concat
+        [ "Invalid $"
+        , stackRootEnvVar
+        , ": '"
+        , envStackRoot
+        , "'"
+        ]
+    show (Won'tCreateStackRootInDirectoryOwnedByDifferentUser envStackRoot parentDir) = concat
+        [ "Preventing creation of $"
+        , stackRootEnvVar
+        , " '"
+        , envStackRoot
+        , "'. Parent directory '"
+        , parentDir
+        , "' is owned by someone else."
+        ]
+    show (UserDoesn'tOwnStackRoot stackRoot) = concat
+        [ "You are not the owner of '"
+        , toFilePath stackRoot
+        , "'. Aborting to protect file permissions."
+        , "\nRetry with '--"
+        , T.unpack configMonoidAllowDifferentUserName
+        , "' to disable this precaution."
+        ]
 instance Exception ConfigException
 
+-- TODO: stackRootEnvVar is defined in Stack.Constants but
+-- can't currently be imported due to a dependency cycle.
+stackRootEnvVar :: String
+stackRootEnvVar = "STACK_ROOT"
```

Comment:
Is there anything that we can do about this? Would there be a point in splitting `Stack.Constants` into a part that is actually constant and a part that depends on other stack modules?